s3store/s3store_test.go

```diff
@@ -24,6 +24,138 @@ var _ tusd.TerminaterDataStore = s3store.S3Store{}
 var _ tusd.FinisherDataStore = s3store.S3Store{}
 var _ tusd.ConcaterDataStore = s3store.S3Store{}
 
+func TestCalcOptimalPartSize(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	assert := assert.New(t)
+
+	s3obj := NewMockS3API(mockCtrl)
+	store := s3store.New("bucket", s3obj)
+
+	assert.Equal("bucket", store.Bucket)
+	assert.Equal(s3obj, store.Service)
+
+	// If you quickly want to override the default values in this test
+	/*
+		store.MinPartSize = 2
+		store.MaxPartSize = 10
+		store.MaxMultipartParts = 20
+		store.MaxObjectSize = 200
+	*/
+
+	debug := false
+	// If you want the results of all tests printed
+	// debug = true
+
+	var MinPartSize = store.MinPartSize
+	var MaxPartSize = store.MaxPartSize
+	var MaxMultipartParts = store.MaxMultipartParts
+	var MaxObjectSize = store.MaxObjectSize
+	var equalparts, lastpartsize int64
+	var err string
+
+	// sanity check
+	if MaxObjectSize > MaxPartSize*MaxMultipartParts {
+		t.Errorf("MaxObjectSize %v can never be achieved, as MaxMultipartParts %v and MaxPartSize %v only allow for an upload of %v bytes total.\n", MaxObjectSize, MaxMultipartParts, MaxPartSize, MaxMultipartParts*MaxPartSize)
+	}
+	var HighestApplicablePartSize int64 = MaxObjectSize / MaxMultipartParts
+	if MaxObjectSize%MaxMultipartParts > 0 {
+		HighestApplicablePartSize++
+	}
+	var RemainderWithHighestApplicablePartSize int64 = MaxObjectSize % HighestApplicablePartSize
+
+	// some of these tests are actually duplicates, as they specify the same size
+	// in bytes - two ways to describe the same thing. That is wanted, in order
+	// to provide a full picture from any angle.
+	testcases := []int64{
```

Comment:
It would be nice if the test cases where separated into situations where a part size can and cannot be calculated, making the result assertion and error checking for straight forward.

Reply:
It is true, that the current set of checks in TestCalcOptimalPartSize are not straight forward.

However, in my opinion the assertions should work with any size you throw at them, no matter, if we expect a success of an error from CalcOptimalPartSize. One reason for that is, that certain testcases might produce the one result with set A of S3 settings but might produce another with set B.

Before we moved the MaxObjectSize condition above the calculation of the optimalPartSize and by that means creating a scenario, where CalcOptimalPartSize could output 0 as a result, the tests were actually very straight forward, because they were linear. No if condition in there. That is the reason, why I preferred optimalPartSize being calculated, even if size > MaxObjectSize.

One could also argue, that the purpose of CalcOptimalPartSize is calculating said part size and checking whether the size of the upload is larger than MaxObjectSize should not even concern this function but actually would belong somewhere else!

E.g. in NewUpload, because why would we create such an upload in the first place, if we could never finish it.

Therefore I would propose to move that condition to NewUpload and keep CalcOptimalPartSize and it's tests clean.

Don't you agree?

Reply:
> However, in my opinion the assertions should work with any size you throw at them, no matter, if we expect a success or an error from CalcOptimalPartSize. 

I don't share this opinion. Test should assert the returned output from a function if you provide it with the specific input, making them straightforward to understand and also debug if necessary. For example, let's take `store.MaxObjectSize + 1`. Should a part size be calculated for this input? I don't know and I cannot tell from this test case. Frankly, I cannot even tell you whether your test verifies that the behavior of CalcPartSize does not change.

>  One reason for that is, that certain testcases might produce the one result with set A of S3 settings but might produce another with set B. Therefore having dividing the test into a set, which is expected to fail, and one, which is expected to succeed, might not actually be realistic.

If you make the test inputs depend on store.MaxPartSize (etc.) then this should not be an issue, as you did.

> Therefore I would propose to move that condition to NewUpload and keep CalcOptimalPartSize and it's tests clean and universal.

Seen that you already took care of this, I welcome this change :)